lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+      convertElementTypeToTensorValueType(
+          SILType::getBuiltinIntegerType(32, context)),
+      ValueOwnershipKind::Owned);
+  // A boolean corresponding to the stayInLoop flag.
+  newHeader->createPHIArgument(convertElementTypeToTensorValueType(
+                                   SILType::getBuiltinIntegerType(1, context)),
+                               ValueOwnershipKind::Trivial);
+
+  // Create a new latch block.
+  SILBasicBlock *latchBlock = currentFn->createBasicBlock();
+  latchBlock->cloneArgumentList(newHeader);
+  {
+    builder.setInsertionPoint(latchBlock);
+    SmallVector<SILValue, 8> latchArgs;
+    for (const SILArgument *latchArg : latchBlock->getArguments()) {
+      latchArgs.push_back(latchArg);
+    }
+    builder.createBranch(
+        getUserSourceLocation(header->getTerminator()->getDebugLocation()),
+        newHeader, latchArgs);
+  }
+
+  llvm::DenseMap<SILBasicBlock *, intmax_t> exitIndices;
+  // As a simplification, simply 0 as exit index for edges to the header
+  // even though header is not an exit block.
+  exitIndices.insert({header, 0});
+  // Identify the exit from the header and also set it to zero index.
+  SILBasicBlock* headerExit = nullptr;
+  for (SILBasicBlock* succ : header->getSuccessorBlocks()) {
+    if (loop->contains(succ))  continue;
+    assert(headerExit == nullptr && "Loop header has more than one exit node.");
+    headerExit = succ;
+  }
+  if (headerExit != nullptr) {
+    exitIndices.insert({headerExit, 0});
+  }
+  unsigned nextExitIndex = 1;
+  for (const auto &edge : edgesToFix) {
+    SILBasicBlock *src = const_cast<SILBasicBlock *>(edge.first);
+    SILBasicBlock *tgt = const_cast<SILBasicBlock *>(edge.second);
+    SILBasicBlock *newTgt = loop->contains(src) ? latchBlock : newHeader;
+    bool stayInLoop = loop->contains(tgt);
+    replaceBranchTarget(src->getTerminator(), tgt, newTgt,
+                        /*preserveArgs=*/stayInLoop);
+    // Set up additional arguments.
+    // If we are exiting the loop, then we should simply use newHeader args.
+    SmallVector<SILValue, 8> newArgs;
+    if (!stayInLoop) {
+      for (unsigned i = 0; i < oldHeaderNumArgs; ++i) {
+        newArgs.push_back(newHeader->getArgument(i));
+      }
+    }
+    SILBuilder builder(src->getTerminator());
+    SILLocation location(
+        getUserSourceLocation(src->getTerminator()->getDebugLocation()));
+    // Find an appropriate value to use for each escaping value.
+    unsigned argIndex = oldHeaderNumArgs;
+    for (const SILValue &escapingValue : escapingValues) {
+      if (DI.properlyDominates(escapingValue, src->getTerminator())) {
+        newArgs.push_back(escapingValue);
+      } else if (src != preheader) {
+        // if src is not the preheader than new header arguments are available.
+        newArgs.push_back(newHeader->getArgument(argIndex));
+      } else {
+        newArgs.push_back(getUndef(escapingValue->getType()));
```

Comment:
i don't fully understand why and how we use undef -- some more context would be nice.

Reply:
Added a brief explanation here.